rl4co/utils/decoding.py

```diff
@@ -0,0 +1,502 @@
+        else:
+            if self.num_starts is not None:
+                if self.num_starts > 1:
+                    log.warn(
+                        f"num_starts={self.num_starts} is ignored for decode_type={self.name}"
+                    )
+
+            self.num_starts = 0
+
+        # Multi-start decoding: first action is chosen by ad-hoc node selection
+        if self.num_starts > 1:
+            if self.select_start_nodes_fn is not None:
+                action = self.select_start_nodes_fn(td, env, self.num_starts)
+            else:
+                action = env.select_start_nodes(td, num_starts=self.num_starts)
+
+            # Expand td to batch_size * num_starts
+            td = batchify(td, self.num_starts)
+
+            td.set("action", action)
+            td = env.step(td)["next"]
+            logprobs = torch.zeros_like(
+                td["action_mask"], device=td.device
+            )  # first logprobs is 0, so p = logprobs.exp() = 1
+
+            self.logprobs.append(logprobs)
+            self.actions.append(action)
+
+        return td, env, self.num_starts
+
+    def post_decoder_hook(
+        self, td: TensorDict, env: RL4COEnvBase
+    ) -> Tuple[torch.Tensor, torch.Tensor, TensorDict, RL4COEnvBase]:
+        assert (
+            len(self.logprobs) > 0
+        ), "No logprobs were collected because all environments were done. Check your initial state"
+
+        return torch.stack(self.logprobs, 1), torch.stack(self.actions, 1), td, env
+
+    def step(
+        self, logits: torch.Tensor, mask: torch.Tensor, td: TensorDict, **kwargs
+    ) -> TensorDict:
+        """Main decoding operation. This method should be called in a loop until all sequences are done.
+        Args:
+            logits: Logits from the model.
+            mask: Action mask. 1 if feasible, 0 otherwise (so we keep if 1 as done in PyTorch).
+            td: TensorDict containing the current state of the environment.
+        """
+        if not self.mask_logits:  # set mask_logit to None if mask_logits is False
+            mask = None
+
+        logprobs = process_logits(
+            logits,
+            mask,
+            temperature=self.temperature,
+            top_p=self.top_p,
+            top_k=self.top_k,
+            tanh_clipping=self.tanh_clipping,
+            mask_logits=self.mask_logits,
+        )
+        logprobs, selected_actions, td = self._step(logprobs, mask, td, **kwargs)
+        td.set("action", selected_actions)
+        self.actions.append(selected_actions)
+        self.logprobs.append(logprobs)
+        return td
+
+    @staticmethod
+    def greedy(logprobs, mask=None):
+        """Select the action with the highest probability."""
+        # [BS], [BS]
+        selected = logprobs.argmax(dim=-1)
+        if mask is not None:
+            assert (
+                not (~mask).gather(1, selected.unsqueeze(-1)).data.any()
+            ), "infeasible action selected"
```

Comment:
Is it still required to check here (also in sampling, L331-337) as the logprobs are already masked in `process_logits` (L298-306)?

Reply:
It can be needed if one passes a custom function that does not mask the logprobs (e.g. if `sampling` and `greedy` are called directly)